bookshelf/api_v3/gce.py

```diff
@@ -0,0 +1,425 @@
+from time import time, sleep
+import uuid
+
+from zope.interface import implementer, provider
+from pyrsistent import PClass, field
+from oauth2client.client import (
+    SignedJwtAssertionCredentials, GoogleCredentials
+)
+from googleapiclient import discovery
+from googleapiclient.errors import HttpError
+
+from bookshelf.api_v2.logging_helpers import log_green, log_yellow, log_red
+from bookshelf.api_v1 import wait_for_ssh
+from cloud_instance import ICloudInstance, ICloudInstanceFactory, Distribution
+
+
+class GCEConfiguration(PClass):
+    credentials_private_key = field(factory=unicode, initial=u"", mandatory=True)
+    credentials_email = field(factory=unicode, initial=u"", mandatory=True)
+    public_key_filename = field(factory=unicode, mandatory=True)
+    private_key_filename = field(factory=unicode, mandatory=True)
+    project = field(factory=unicode, mandatory=True)
+    machine_type = field(factory=unicode, mandatory=True)
+    image_basename = field(type=unicode, mandatory=True, factory=unicode)
+    username = field(factory=unicode, mandatory=True)
+    description = field(factory=unicode, mandatory=True)
+    instance_name = field(factory=unicode, mandatory=True)
+    base_image_prefix = field(factory=unicode, mandatory=True)
+    base_image_project = field(factory=unicode, mandatory=True)
+
+
+class GCEState(PClass):
+    instance_name = field(factory=unicode, mandatory=True)
+    ip_address = field(factory=unicode, mandatory=True)
+    distro = field(factory=unicode, mandatory=True)
+    zone = field(factory=unicode, mandatory=True)
+
+
+@implementer(ICloudInstance)
+@provider(ICloudInstanceFactory)
+class GCE(object):
+
+    cloud_type = 'gce'
+
+    def __init__(self, config, state):
+        self.config = GCEConfiguration.create(config)
+        self.state = state
+        self._compute = self._get_gce_compute()
+
+    @property
+    def project(self):
+        return self.config.project
+
+    @property
+    def username(self):
+        return self.config.username
+
+    @property
+    def zone(self):
+        return self.state.zone
+
+    @property
+    def region(self):
+        return self.state.zone
+
+    @property
+    def distro(self):
+        return Distribution(self.state.distro)
+
+    @property
+    def description(self):
+        return self.config.description
+
+    @property
+    def image_basename(self):
+        return self.config.image_basename
+
+    @property
+    def ip_address(self):
+        return self.state.ip_address
+
+    @property
+    def key_filename(self):
+        return self.config.private_key_filename
+
+    @classmethod
+    def create_from_config(cls, config, distro, region):
+        instance_name = "{}-{}".format(
```

Comment:
The docstring will be the same as the docstring on the interface (`cloud_instance.ICloudInstanceFactory`). Lemmie know if that doesn't suffice.